src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -129,6 +133,71 @@ split_change_cb(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
 	}
 
 	Relation targetRelation = RelationIdGetRelation(targetRelationOid);
+
+	/*
+	 * If any columns from source relation have been dropped, then the tuple needs to
+	 * be formatted according to the target relation.
+	 */
+	TupleDesc sourceRelationDesc = RelationGetDescr(relation);
+	TupleDesc targetRelationDesc = RelationGetDescr(targetRelation);
+	if (sourceRelationDesc->natts > targetRelationDesc->natts)
+	{
+		switch (change->action)
+		{
+			case REORDER_BUFFER_CHANGE_INSERT:
+			{
+				HeapTuple sourceRelationNewTuple = &(change->data.tp.newtuple->tuple);
+				HeapTuple targetRelationNewTuple = GetTupleForTargetSchema(
+					sourceRelationNewTuple, sourceRelationDesc, targetRelationDesc);
+
+				change->data.tp.newtuple->tuple = *targetRelationNewTuple;
+				break;
+			}
+
+			case REORDER_BUFFER_CHANGE_UPDATE:
+			{
+				HeapTuple sourceRelationNewTuple = &(change->data.tp.newtuple->tuple);
+				HeapTuple targetRelationNewTuple = GetTupleForTargetSchema(
+					sourceRelationNewTuple, sourceRelationDesc, targetRelationDesc);
+
+				change->data.tp.newtuple->tuple = *targetRelationNewTuple;
+
+				/*
+				 * For a relation with REPLICA_IDENTITY_FULL, we need to format the old tuple too
+				 * according to the target relation.
+				 */
+				if (relation->rd_rel->relreplident == REPLICA_IDENTITY_FULL ||
```

Comment:
is this check right? As you pointed out the old tuple also gets all the columns in ExtractReplicaIdentity, and the main difference in heap_update seems to be that we don't log the oldtuple if the column has not changed.

In that case, should this be `if (change->data.tp.oldtuple != NULL)` ?

Reply:
@marcocitus : Thanks for the explanation! 
I have fixed this condition. Also added one permutation where we have a replica idenity with two columns(non-full) and one column amongst it is updated to exercise this `if`